### PR TITLE
padding nanosec-timestamp with zeroes

### PR DESCRIPTION
**Pull request by ronish31**
@AmirTugi 
Please review change:
Fixing issue with sorting logs by nanosec-timestamp by padding the nano-timestamp with zeros.
1de20de1f9a92351ef9f0e75afa69f86f1f2aa12

### CODE REVIEW

**Review by AmirTugi — CHANGES_REQUESTED** [review]
* Use [`padStart`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart)/[`padEnd`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padEnd)
* Please add a test for this :D

**Comment by ronish31** [comment]
@AmirTugi will do

**Comment by ronish31** [comment]
**fixed comment and added tests: ** https://github.com/logzio/logzio-nodejs/pull/79/commits/402ab5752aec36cb8fdc0487d7e1e3ea29093e84
@AmirTugi

File: test/logger.test.js [hunk 1]
```diff
@@ -191,14 +195,52 @@ describe('logger', () => {
             });
             sinon.spy(logger, '_createBulk');
 
+            logger.log('hello there from test');
```
Comment (AmirTugi) [inline test/logger.test.js]: This `it` has 2 different tests
* Not adding nano seconds
* Adding nano seconds
Please split it

Comment (AmirTugi) [reply test/logger.test.js]: The string should be a var so you wouldn't have to maintain it in multiple places

File: test/logger.test.js [hunk 2]
```diff
@@ -191,14 +195,52 @@ describe('logger', () => {
...
+            };
+            logger._addTimestamp(mock_log);
+            let mock_nano_timestamp = mock_log['@timestamp_nano'].split("-")[nanosec_index];
+
+            assert.equal(mock_nano_timestamp.length, nanoSecDigits);
+
+            logger.close();
+        });
+
+        it('pad nano-sec timestamp with zeros', (done) => {
```
Comment (AmirTugi) [inline test/logger.test.js]: Please merge both tests.
What the test should do is to expect the timestamp to be something you'd know (like `expectedTimestamp`) and then see that that was indeed the timestamp of the log.
You'd probably have to create relevant mocks.

**Review by AmirTugi — CHANGES_REQUESTED** [review]
The test should be a minimal as possible, and check as much value as possible.
In this case, you'd want to ask for nano sec timestamp, and mock the time to be able to expect the correct nano sec timestamp.

File: lib/logzio-nodejs.js [hunk 3]
```diff
@@ -184,7 +185,7 @@ class LogzioLogger {
 
         if (this.addTimestampWithNanoSecs) {
             const time = process.hrtime();
-            msg['@timestamp_nano'] = msg['@timestamp_nano'] || [now, time[0].toString(), time[1].toString()].join('-');
+            msg['@timestamp_nano'] = msg['@timestamp_nano'] || [now, this._padNumberWithZeros(time[1].toString())].join('-');
```
Comment (AmirTugi) [inline lib/logzio-nodejs.js]: Why not just
```suggestion
            msg['@timestamp_nano'] = msg['@timestamp_nano'] || [now, time[1].toString().padStart(nanoSecDigit, '0')].join('-');
```

Comment (ronish31) [reply lib/logzio-nodejs.js]: Divided to 3 tests: 
default case of no nanosec-timestamp
add nano-sec timestamp option
valid nano-sec timestamp
https://github.com/logzio/logzio-nodejs/pull/79/commits/c326c92c109a70e751218df3a9f8798ba2f03399
@AmirTugi

Comment (ronish31) [reply lib/logzio-nodejs.js]: fixed

File: test/logger.test.js (hunk 2 above)
Comment (ronish31) [reply test/logger.test.js]: Divided to 3 tests:
default case of no nanosec-timestamp
add nano-sec timestamp option
valid nano-sec timestamp
c326c92
@AmirTugi

File: package.json [hunk 4]
```diff
@@ -37,6 +37,7 @@
     "logzio"
   ],
   "dependencies": {
+    "hrtimemock": "^2.0.0",
```
Comment (AmirTugi) [inline package.json]: That's a mock package for testing.
Please put it in `devDependencies`

File: test/logger.test.js [hunk 5]
```diff
@@ -168,10 +173,10 @@ describe('logger', () => {
             logger.close();
         });
 
-        it('adds nano seconds when added to options', (done) => {
-            // testing without nano seconds
+        it('default case of no nanosec-timestamp', (done) => {
```
Comment (AmirTugi) [inline test/logger.test.js]: Please say what the test does in the format of
`it('should ...')`
e.g - `it('should not include nano timestamp by default')`

File: test/logger.test.js [hunk 6]
```diff
@@ -182,8 +187,9 @@ describe('logger', () => {
 
             logger._createBulk.restore();
             logger.close();
+        });
 
-            // testing with nano seconds
+        it('add nano-sec timestamp option', (done) => {
```
Comment (AmirTugi) [inline test/logger.test.js]: As I said, this test is redundant.
You're already checking this use case in the next test.

File: test/logger.test.js [hunk 7]
```diff
@@ -192,13 +198,36 @@ describe('logger', () => {
...
+
+        it('valid nano-sec timestamp', (done) => {
+            const nanosec_index = 3;
+            logger = createLogger({
+                bufferSize: 1,
+                callback: done,
+                addTimestampWithNanoSecs: true,
+            });
+            sinon.spy(logger, '_createBulk');
+            hrtimemock(0.123456);
```
Comment (AmirTugi) [inline test/logger.test.js]: For more readability please put the mock time and the expected test time to vars called `mockTime` and `expectedLogTime`

File: test/logger.test.js [hunk 8]
```diff
@@ -192,13 +198,36 @@ describe('logger', () => {
...
+            sinon.spy(logger, '_createBulk');
+            hrtimemock(0.123456);            
+            process.hrtime();
+            logger.log({
+                message: 'hello there from test'
+            })
+            const mock_log = logger._createBulk.getCall(0).args[0][0];
+            const mock_nano_timestamp = mock_log['@timestamp_nano'].split("-")[nanosec_index];
+
+            assert.equal(mock_nano_timestamp, '000123456');
```
Comment (AmirTugi) [inline test/logger.test.js]: Just check the truthiness of `mock_log['@timestamp_nano'].endsWith('<theExpectedValue>')`
Check if there's an assert with `endsWith`

File: test/logger.test.js [hunk 9]
```diff
@@ -5,9 +5,13 @@ const assert = require('assert');
 const moment = require('moment');
 const zlib = require('zlib');
 const logzioLogger = require('../lib/logzio-nodejs.js');
+var hrtimemock = require('hrtimemock');
```
Comment (AmirTugi) [inline test/logger.test.js]:
```suggestion
const hrtimemock = require('hrtimemock');
```

File: test/logger.test.js [hunk 10]
```diff
@@ -5,9 +5,13 @@ const assert = require('assert');
...
 
 const dummyHost = 'logz.io';
 const nockHttpAddress = `http://${dummyHost}:8070`;
+const nanosecAsDecimal = function nanosecAsDecimal(nanosec_timestamp){
```
Comment (AmirTugi) [inline test/logger.test.js]: You're not using this

**Review by AmirTugi — CHANGES_REQUESTED** [review]
Very nice.
Almost good to go :)

File: test/logger.test.js (hunk 8 above)
Comment (ronish31) [reply test/logger.test.js]: fixed with:
 assert.equal(mockLog['@timestamp_nano'].endsWith(expectedLogTime), true);

File: test/logger.test.js [hunk 11]
```diff
@@ -182,23 +184,30 @@ describe('logger', () => {
 
             logger._createBulk.restore();
             logger.close();
+        });
+
+        it('valid nano-sec timestamp', (done) => {
```
Comment (AmirTugi) [inline test/logger.test.js]:
```suggestion
        it('should add a nano-sec timestamp to the log', (done) => {
```

File: test/logger.test.js [hunk 12]
```diff
@@ -182,23 +184,30 @@ describe('logger', () => {
...
-
+            hrtimemock(mockTime);            
+            process.hrtime();
             logger.log({
-                message: 'hello there from test',
-            });
-            assert.equal(logger._createBulk.getCall(0).args[0][0].hasOwnProperty('@timestamp_nano'), true);
+                message: 'hello there from test'
+            })
+            const mockLog = logger._createBulk.getCall(0).args[0][0];
```
Comment (AmirTugi) [inline test/logger.test.js]:
```suggestion
            const mockLogCall = logger._createBulk.getCall(0).args[0][0];
```

File: test/logger.test.js [hunk 13]
```diff
@@ -182,23 +184,30 @@ describe('logger', () => {
...
-            // testing with nano seconds
             logger = createLogger({
                 bufferSize: 1,
                 callback: done,
                 addTimestampWithNanoSecs: true,
             });
             sinon.spy(logger, '_createBulk');
-
+            hrtimemock(mockTime);            
+            process.hrtime();
```
Comment (AmirTugi) [inline test/logger.test.js:200]: Why do you need this?
Remove if unnecessary

**Review by AmirTugi — APPROVED** [review]

File: test/logger.test.js (hunk 13 above)
Comment (ronish31) [reply test/logger.test.js:200]: That's the nanosec-timestamp mock - it sets the _addtimestamp function to get a specific nanosec time.

**Pull request merged**